src/rust/engine/process_execution/bazel_protos/build.rs

```diff
@@ -1,29 +1,64 @@
-use std::env::home_dir;
-use std::process::{exit, Command};
+extern crate build_utils;
+extern crate protoc_grpcio;
+
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+use build_utils::BuildRoot;
 
 fn main() {
-  println!("cargo:rerun-if-changed=../../../../../3rdparty/protobuf");
-  println!("cargo:rerun-if-changed=generate-grpc.sh");
-
-  let cachedir = home_dir()
-    .unwrap()
-    .join(".cache")
-    .join("pants")
-    .join("rust")
-    .join("cargo")
-    .join("bin");
-  println!(
-    "cargo:rerun-if-changed={}",
-    cachedir.join("grpc_rust_plugin").to_str().unwrap()
-  );
+  let build_root = BuildRoot::find().unwrap();
+  let thirdpartyprotobuf = build_root.join("3rdparty/protobuf");
   println!(
     "cargo:rerun-if-changed={}",
-    cachedir.join("protoc-gen-rust").to_str().unwrap()
+    thirdpartyprotobuf.to_str().unwrap()
   );
 
-  let output = Command::new("./generate-grpc.sh").output().unwrap();
+  let gen_dir = PathBuf::from("src/gen");
+
+  // Re-gen if, say, someone does a git clean on the gen dir but not the target dir. This ensures
+  // generated sources are available for reading by programmers and tools like rustfmt alike.
+  println!("cargo:rerun-if-changed={}", gen_dir.to_str().unwrap());
+
+  protoc_grpcio::compile_grpc_protos(
+    &[
+      "google/devtools/remoteexecution/v1test/remote_execution.proto",
+      "google/bytestream/bytestream.proto",
+      "google/rpc/code.proto",
+      "google/rpc/error_details.proto",
+      "google/rpc/status.proto",
+      "google/longrunning/operations.proto",
+      "google/protobuf/empty.proto",
+    ],
+    &[
+      thirdpartyprotobuf.join("googleapis"),
+      thirdpartyprotobuf.join("standard"),
+      thirdpartyprotobuf.join("rust-protobuf"),
+    ],
+    &gen_dir,
+  ).expect("Failed to compile protos!");
+
+  let listing = gen_dir.read_dir().unwrap();
+  let mut pub_mod_stmts = listing
+    .filter_map(|d| {
+      let dirent = d.unwrap();
+      let file_name = dirent.file_name().into_string().unwrap();
+      match file_name.trim_right_matches(".rs") {
+        "mod" | ".gitignore" => None,
+        module_name => Some(format!("pub mod {};", module_name)),
+      }
+    })
+    .collect::<Vec<_>>();
+  pub_mod_stmts.sort();
+  let contents = format!(
+    "\
+// This file is generated. Do not edit.
+{}
+",
+    pub_mod_stmts.join("\n")
+  );
 
-  print!("{}", String::from_utf8(output.stdout).unwrap());
-  eprint!("{}", String::from_utf8(output.stderr).unwrap());
-  exit(output.status.code().unwrap());
+  let mut module = File::create(gen_dir.join("mod.rs")).unwrap();
```

Comment:
In production code I'd probably write this as:
```
File::create(gen_dir.join("mod.rs"))
  .and_then(|file| file.write_all(contents.as_bytes()))
  .expect("Failed to write mod.rs")
```

both because I don't care whether the create or write failed, and to give a useful error on failure, but in a build.rs I care less :)

Reply:
I like it, done.